2019-03-Autumn/3_stacks_queues/README.md

```diff
@@ -102,4 +120,89 @@ TODO :bug:
 
 TODO :bug:
 
-
+### 3. Reverse Substrings In Parentheses
+
+Source: [LeetCode](https://leetcode.com/problems/reverse-substrings-between-each-pair-of-parentheses/) 
+
+#### Solution Approach
+
+The essential problem here is to determine the order of strings to 
+reverse, since we expect arbitrary parenthesis pairs to be nested 
+within each other. The nature of this problem lends itself to stacks 
+very well.
+
+If you create a stack that represents instances of opening parentheses 
+(by, for example, recording the index location of that character), you 
+will be able to pop the most recent index location when you find a 
+closing parenthesis. 
+
+When you have this index pair, you can reverse the contained substring 
+and continue searching for the next parenthesis (open or closed).
+
+The remaining problem we have is how to ensure the contained string 
+is updated properly so that higher levels of nested strings reverse the 
+string properly.The solution provided here updates the string in-place 
+and manages the traversal index manually to account for removed 
+parentheses. Other implementations for this subproblem exist.
+
+_Note: the given [constraints](#constraints) are particularly useful 
+for this problem statement and remove a lot of the edge case checking 
+we would need for a more robust solution._
+
+#### Solution
+
+The C++ solution in its entirety is:
+
+```c++
+// Solution function.
+string reverseParentheses(string s) {
+    // Create a stack to store the index position of all open parentheses.
+    stack<int> open; 
+    for (int i = 0; i < s.length(); i++) {
+        // We have reached a new subset for reversal.
+        // Add this index to the stack.
+        if (s[i] == '(') {
+            open.push(i);
+        // We have finished traversing a subset to be reversed; 
+        // now, reverse it.
+        } else if (s[i] == ')') {
+            stringstream n;
+            int start = open.top();
+            open.pop();
+            // Use utility function to reverse found substring.
+            // NOTE: here, we intentionally exclude the parentheses.
+            n << reverseString(s.substr(start + 1, i - start - 1));
+            // We then modify the string in-place (replacing 
+            // parentheses grouping with reversed inner string),
+            // and update our index value to account for the removed 
+            // characters.
+            s.replace(start, i - start + 1, n.str());
+            i -= 2;
+        }
+    }
+    return s;
+}
+
+// Helper function.
+string reverseString(string r) {
+    stringstream s;
+    for (int i = r.length() - 1; i >= 0; i--) {
+        s << r[i];
+    }
+    return s.str();
```

Comment:
Don't really need our own reverseString when there's `std::reverse`

Reply:
I am leaving it in for now; it isn't especially complicated and it reduces some pressure to know the standard libraries inside and out. I'm adding a comment to that effect.